From a22ee8fb0d135dc1ad89c1fc85c30b22016cb8a6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 22 Feb 2022 11:23:19 +0100
Subject: [PATCH 30/35] winegstreamer: Implement wg_pool_alloc_buffer and
 wg_pool_free_buffer.

For: Call of Duty III, Mortal Kombat 11, Shadow Warrior 2,
Yakuza 4 Remastered, Hard Reset Redux.

Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=45988
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=47084
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=49715
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=52183
CW-Bug-Id: #16839
CW-Bug-Id: #18678
CW-Bug-Id: #19362
---
 dlls/winegstreamer/wg_pool.c | 21 ++++++++++++++++++---
 1 file changed, 18 insertions(+), 3 deletions(-)

diff --git a/dlls/winegstreamer/wg_pool.c b/dlls/winegstreamer/wg_pool.c
index 11111111111..11111111111 100644
--- a/dlls/winegstreamer/wg_pool.c
+++ b/dlls/winegstreamer/wg_pool.c
@@ -59,17 +59,32 @@ static GstFlowReturn wg_pool_alloc_buffer(GstBufferPool *gst_pool, GstBuffer **b
         GstBufferPoolAcquireParams *params)
 {
     WgBufferPool *pool = (WgBufferPool *)gst_pool;
+    GstVideoInfo *video_info = pool->is_video ? &pool->u.video_info : NULL;
 
-    GST_FIXME("pool %p, buffer %p, params %p, stub!", pool, buffer, params);
+    GST_DEBUG("pool %p, buffer %p, params %p", pool, buffer, params);
 
-    return GST_FLOW_ERROR;
+    *buffer = gst_buffer_new_and_alloc(pool->buffer_size);
+
+    if (*buffer && video_info)
+    {
+        gst_buffer_add_video_meta_full(*buffer, GST_VIDEO_FRAME_FLAG_NONE,
+                GST_VIDEO_INFO_FORMAT(video_info), GST_VIDEO_INFO_WIDTH(video_info),
+                GST_VIDEO_INFO_HEIGHT(video_info), GST_VIDEO_INFO_N_PLANES(video_info),
+                video_info->offset, video_info->stride);
+    }
+
+    GST_INFO("Returning buffer %p", *buffer);
+    return *buffer ? GST_FLOW_OK : GST_FLOW_ERROR;
 }
 
 static void wg_pool_free_buffer(GstBufferPool *gst_pool, GstBuffer *buffer)
 {
     WgBufferPool *pool = (WgBufferPool *)gst_pool;
 
-    GST_FIXME("pool %p, buffer %p, stub!", pool, buffer);
+    GST_DEBUG("pool %p, buffer %p", pool, buffer);
+
+    /* release the last ref on the buffer */
+    gst_buffer_unref(buffer);
 }
 
 static void wg_pool_init(WgBufferPool *pool)

